blivet/devicefactory.py

```diff
@@ -1752,6 +1759,92 @@ def _get_new_device(self, *args, **kwargs):
         return super(LVMThinPFactory, self)._get_new_device(*args, **kwargs)
 
 
+class LVMVDOFactory(LVMFactory):
+
+    """ Factory for creating LVM VDO volumes.
+    """
+
+    def __init__(self, storage, **kwargs):
+        self.pool_name = kwargs.pop("pool_name", None)
+        self.virtual_size = kwargs.pop("virtual_size", None)
+        self.compression = kwargs.pop("compression", False)
+        self.deduplication = kwargs.pop("deduplication", False)
+        super(LVMVDOFactory, self).__init__(storage, **kwargs)
+
+    def _get_new_pool(self, *args, **kwargs):
+        kwargs["vdo_pool"] = True
+        return super(LVMVDOFactory, self)._get_new_device(*args, **kwargs)
+
+    def _set_device_size(self):
+        """ Set the size of the factory device. """
+        super(LVMVDOFactory, self)._set_device_size()
+
+        self.device.pool.size = self.size
+        self._reconfigure_container()
+
+        if not self.virtual_size or self.virtual_size < self.size:
+            # virtual_size is not set or smaller than current size --> it should be same as the pool size
+            self.device.size = self.size
+        else:
+            self.device.size = self.virtual_size
+
+    def _set_pool_name(self):
+        safe_new_name = self.storage.safe_device_name(self.pool_name,
+                                                      get_device_type(self.device))
+        if self.device.pool.name != safe_new_name:
+            if not safe_new_name:
+                log.error("not renaming '%s' to invalid name '%s'",
+                          self.device.pool.name, self.pool_name)
+                return
+            if safe_new_name in self.storage.names:
+                log.error("not renaming '%s' to in-use name '%s'",
+                          self.device.pool.name, safe_new_name)
+                return
+
+            log.debug("renaming device '%s' to '%s'",
+                      self.device.pool.name, safe_new_name)
+            self.device.pool.raw_device.name = safe_new_name
+
+    def _set_name(self):
+        super(LVMVDOFactory, self)._set_name()
+        if self.pool_name:
+            self._set_pool_name()
+
+    def _reconfigure_device(self):
+        super(LVMVDOFactory, self)._reconfigure_device()
+
+        self.device.pool.compression = self.compression
+        self.device.pool.deduplication = self.deduplication
```

Comment:
No `virtual_size` reconfiguration?

Reply:
That's covered in `_set_device_size` which is called from the generic `_reconfigure_device`.